fontdrasil/src/orchestration.rs

```diff
@@ -0,0 +1,102 @@
+//! Common parts of work orchestration.
+
+use std::{
+    collections::HashSet,
+    fmt::Debug,
+    hash::Hash,
+    sync::{Arc, RwLock},
+};
+
+pub const MISSING_DATA: &str = "Missing data, dependency management failed us?";
+
+/// A unit of work safe to run in parallel
+///
+/// Naively you'd think we'd just return FnOnce + Send but that didn't want to compile
+/// See <https://github.com/rust-lang/rust/issues/29625>.
+///
+/// Data produced by work is written into a Context (type parameter C).
+pub trait Work<C, E> {
+    fn exec(&self, context: &C) -> Result<(), E>;
+}
+
+#[derive(Clone)]
+pub struct Cache<T: Default> {
+    pub item: Arc<RwLock<T>>,
+}
+
+impl<T: Default> Cache<T> {
+    pub fn new() -> Cache<T> {
+        Cache {
+            item: Default::default(),
+        }
+    }
+}
+
+impl<T: Default> Default for Cache<T> {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+/// Tracks what parts of a context we can read/write.
+///
+/// Not meant to prevent malicious access, merely to detect mistakes
+/// as the result of mistaken concurrent access can be confusing to track down.
+pub struct Acl<I>
+where
+    I: Eq + Hash + Debug,
+{
+    // If present, the one and only key you are allowed to write to
+    // None means we're read-only
+    write_mask: Option<I>,
+
+    // If present, what you can access through this context
+    // Intent is root has None, task-specific Context only allows access to dependencies
+    read_mask: Option<HashSet<I>>,
+}
```

Comment:
Looking at this makes me wonder if we want to be an enum, with `Read` and `ReadWrite` variants or something?

like,

```rust
enum Access {
    Read { read_mask: Option<HashSet<I>> },
    ReadWrite { read_mask: Option<HashSet<I>>, write_mask: I },
}
```

And to keep going, possibly we should have a type, `ReadMask`, like:

```rust
enum ReadMask {
    Root,
    Items(HashSet<I>),
}
```

but maybe not, I'm just spitballing.

Reply:
I think it works as a single set of enum Access Read(I) Write(I). Let me try that and see if it feels better.

Reply:
It looked lovely but led me to copy the &I because I wanted to lookup an &I in a HashSet<Access<I>>. Might just be a rust noob mistake, I'll make a gist and send to chat. For the purpose of this PR I will leave as is and then update to Something Better later.

Reply:
https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=07b82a0988efb3869e40e2be855c2354